README.md
Final Project of Modern Data Struture (GR5072)
=============
community_ry2403
------------

A Python package for aggregating community-level Census Data.

Install from: https://test.pypi.org/project/community-ry2403/


Introduction
------------

This library provides a Python interface for the [ACS API](https://www.census.gov/data/developers/data-sets/acs-5year.html), [CBP API](https://www.census.gov/data/developers/data-sets/cbp-nonemp-zbp/cbp-api.html)  and [ZCBP API](https://www.census.gov/data/developers/data-sets/cbp-nonemp-zbp/zbp-api.html).      


ACS API data are reported on ZIP Code Tabulation Areas(ZCTA) level while others are reported on Zip Code level.      

The package allows users to get data from API and provides some tools for converting the report level.   


Methology
------------
How to convert?
* For census data (from ACS)
    - If report level is ZCTA, keep the original value.
    - If report level is ZIP, 
        - for percent data, use the original value within the correspondent ZCTA area, 
        - for absolute number, use the mean estimate within correspondent ZCTA area.

* For business data (from CBP,ZCBP)
    - If report level is ZIP, keep the original value.
    - If report level is ZCTA, 
        - for percent data, use the mean estimate within the correspondent ZIP areas, 
        - for absolute number, use the sum estimate within correspondent ZIP areas.


Installing
----------

It works with Python versions from 3.9+.

You can install using:

```shell
    pip install -i https://test.pypi.org/simple/ community-ry2403
```



Running Tests
-------------

```shell
poetry add --dev pytest
poetry run pytest
```



Basic Tutorial 
--------------


### search possible variables

```python
from community_ry2403.community_ry2403 import variables
#Initialize the class variables
f = variables()   
f.find_variable(keyword='total households by marry')
```

### Get the crosswalk of ZCTA-ZIP-State

```python
from community_ry2403.community_ry2403 import get_code
get_code(area_code=['10025','10036'],level='zip')
```

### Query data from ACS & CBP & ZCBP API

```python
from community_ry2403.community_ry2403 import census_data        
from community_ry2403.community_ry2403 import business_data        

v1=['DP02_0001E','DP02_0002E','DP02_0003PE']
census_data(year=2019,variable=v1,area_code=['10025','10036'])       

v2 = ['EMP','EMP_N','ESTAB','PAYANN','PAYANN_N','PAYQTR1']
business_data(year=2019,variable=v2,area_code=['10025','10036'],industry=72)

```

### Query data with customized report level
```python
from community_ry2403.community_ry2403 import search
# Initialize the api
api = search()  
v=['DP02_0001E','DP02_0002E','DP02_0003PE']
api.census(area_code=['10025','10036'],geography='zip',year=2019, variable=v)
```



community_ry2403/CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.0 (12/12/2021)

- First release of `community_ry2403`!

community_ry2403/tests/test_community_ry2403.py
from community_ry2403 import community_ry2403

import pytest 
import pandas as pd

def test_search():

    area_code = ['79925','10025']
    geography = 'Zip'
    year=2019
    variable = ['ESTAB','EMP_N','PAYQTR1_N']
    industry=72
    
    api = community_ry2403.search()
    actual = api.business(area_code, geography ,year, variable,industry)

    assert type(actual)==pd.DataFrame,'test failed'


def test_variables():
    find = community_ry2403.variables()
    actual = find.find_variable('total households marry people percent')
    assert type(actual)==pd.DataFrame,'test failed'


community_ry2403/_build/searchindex.js
Search.setIndex({docnames:["autoapi/community_ry2403/community_ry2403/index","autoapi/community_ry2403/index","autoapi/index","changelog","conduct","contributing","example","index"],envversion:{"sphinx.domains.c":2,"sphinx.domains.changeset":1,"sphinx.domains.citation":1,"sphinx.domains.cpp":4,"sphinx.domains.index":1,"sphinx.domains.javascript":2,"sphinx.domains.math":2,"sphinx.domains.python":3,"sphinx.domains.rst":2,"sphinx.domains.std":2,"sphinx.ext.viewcode":1,sphinx:56},filenames:["autoapi\\community_ry2403\\community_ry2403\\index.rst","autoapi\\community_ry2403\\index.rst","autoapi\\index.rst","changelog.md","conduct.md","contributing.md","example.ipynb","index.md"],objects:{"":[[1,0,0,"-","community_ry2403"]],"community_ry2403.community_ry2403":[[0,2,1,"","business_data"],[0,2,1,"","census_data"],[0,1,1,"","codebook"],[0,1,1,"","codebook_path"],[0,2,1,"","get_code"],[0,2,1,"","get_ratio"],[0,1,1,"","label_list"],[0,1,1,"","label_path"],[0,3,1,"","search"],[0,3,1,"","variables"]],"community_ry2403.community_ry2403.search":[[0,4,1,"","business"],[0,4,1,"","census"]],"community_ry2403.community_ry2403.variables":[[0,4,1,"","find_variable"]],community_ry2403:[[1,1,1,"","__version__"],[0,0,0,"-","community_ry2403"]]},objnames:{"0":["py","module","Python module"],"1":["py","data","Python data"],"2":["py","function","Python function"],"3":["py","class","Python class"],"4":["py","method","Python method"]},objtypes:{"0":"py:module","1":"py:data","2":"py:function","3":"py:class","4":"py:method"},terms:{"0":[0,6],"1":[0,2,4,6],"10":0,"10025":0,"10036":0,"11":[0,6],"13785":0,"17260":0,"2":0,"2017":0,"2019":0,"214":0,"255":0,"3":0,"4":[0,4],"41355":0,"5":0,"6":0,"7":0,"72":0,"79925":0,"8":0,"9":[0,6],"case":0,"do":4,"function":5,"import":6,"int":0,"public":4,"return":0,A:[0,7],BY:0,Being:4,By:[5,7],If:[0,5],In:4,It:7,The:[4,5],To:6,__version__:[1,6],abid:[5,7],about:5,absolut:0,abus:4,ac:0,accept:4,account:4,act:4,action:4,adapt:4,add:5,addit:5,address:4,advanc:4,ag:4,aggerag:7,agre:[5,7],align:4,all:[4,5],alvai:6,alwai:5,american:0,an:4,ani:[4,5],anyth:5,api:0,appdata:6,appear:4,appli:4,appoint:4,appreci:5,appropri:[4,5],ar:[4,5],area:0,area_cod:0,articl:5,attack:4,attent:4,auto:2,autoapi:2,avail:0,b:5,ban:4,base:0,befor:5,behavior:4,best:4,bit:5,blog:5,bodi:4,both:4,branch:5,bugfix:5,busi:0,business_data:0,businiss:0,c:6,cach:6,can:5,cbp:0,censu:[0,7],census_data:0,chang:5,check:[5,7],checkout:5,circumst:4,clarifi:4,code:[0,7],codebook:0,codebook_path:0,comment:4,commit:[4,5],commun:[0,4,6,7],community_ry2403:[2,3,5,6],complaint:4,conduct:7,confidenti:4,conform:5,consid:4,construct:4,contact:4,contain:[0,2],contribut:4,contributor:4,cookiecutt:7,copi:5,correct:4,correspond:0,could:4,coven:4,creat:[2,4,5,7],credit:5,critic:4,current:5,custom:0,data:[0,7],datafram:0,deem:4,defin:4,derogatori:4,detail:[4,5],determin:4,develop:5,differ:4,disabl:4,doc:5,docstr:5,document:2,done:5,download:5,dp02_0001e:0,dp02_0001e_zip:0,dp02_0001pe:0,dp02_0001pe_zip:0,dp02_0014e:0,dp02_0014pe:0,dp02_0015e:0,dp02_0015pe:0,dp02pr_0001e:0,dp02pr_0001pe:0,dp02pr_0014e:0,dp02pr_0014pe:0,dp02pr_0015e:0,dp02pr_0015pe:0,driven:5,e:4,easier:5,edit:4,electron:4,emp_n:0,emp_n_zcta:0,empathi:4,enhanc:5,enough:5,environ:4,estab:0,estab_zcta:0,estim:0,ethnic:4,even:5,event:4,everi:5,everyon:4,exampl:[0,4],expect:4,experi:4,explain:5,explicit:4,express:4,face:4,fair:4,faith:4,feel:5,find:0,find_vari:0,first:3,focus:4,follow:4,format:5,foster:4,free:[4,5],from:[0,4],further:4,fuzz:6,fuzzywuzzi:6,gender:4,gener:2,geographi:0,get:0,get_cod:0,get_ratio:0,git:5,github:5,given:5,good:4,gracefulli:4,greatli:5,guidelin:7,gz4twsim:6,harass:4,harm:4,have:[4,5],help:5,here:5,homepag:4,househol:0,household:0,how:5,ident:4,identifi:0,imageri:4,inappropri:4,incid:4,includ:[4,5],inclus:4,individu:4,industri:0,inform:4,insensit:0,instal:[5,6],instanc:4,insult:4,intereset:0,interest:[4,7],investig:4,issu:[4,5],its:[4,5,7],keep:[0,5],keyword:0,label:0,label_list:0,label_path:0,languag:4,leadership:4,level:[0,4,7],levenshtein:6,lib:6,list:0,littl:5,local:[5,6],look:[0,5],mai:[0,4],mail:4,maintain:4,make:[4,5],marri:0,mean:0,media:4,meet:5,member:4,might:5,mit:7,naic:0,name:5,narrow:5,nation:4,never:5,note:[5,7],number:0,object:0,oblig:4,offens:4,offici:[4,5],offlin:4,onlin:4,open:[4,5],oper:5,orient:4,origin:0,other:4,otherwis:4,out:[0,7],packag:[6,7],page:2,panda:0,paramet:0,part:5,particip:4,pass:5,pattern:0,payqtr1_n:0,payqtr1_n_zcta:0,peopl:0,percent:0,perman:4,permiss:4,person:4,physic:4,pip:7,pkg:7,pleas:[5,7],poetri:5,polici:4,polit:4,posit:4,possibl:5,post:[4,5],print:6,privat:4,profession:4,project:[4,5,6,7],propos:5,publish:4,pure:6,py3:6,py:[6,7],pypoetri:6,python:[5,6,7],race:4,re:5,readi:5,reason:4,regard:4,regardless:4,reject:4,releas:[3,5,7],religion:4,rememb:5,remov:[4,6],renfang:7,repercuss:4,report:[0,4],repres:4,represent:4,reproduc:5,requir:5,respect:4,respond:4,review:4,right:4,ry2403:6,s:[4,5],scope:5,search:0,self:0,separ:4,sequencematch:6,set:[4,5],setup:5,sexual:4,should:5,show:4,similar:5,site:6,size:4,slow:6,social:4,sourc:0,space:4,specif:4,sphinx:2,step:5,str1:0,str2:0,str:0,string:0,sum:0,support:5,survei:0,system:5,tag:5,take:4,team:4,templat:7,temporari:4,temporarili:4,term:[5,7],test:5,thei:[4,5],thi:[2,4,5,6,7],threaten:4,through:5,todo:7,total:0,toward:4,troll:4,troubleshoot:5,type:0,unaccept:4,under:7,unwelcom:4,up:[0,5],updat:5,us:[0,4,5,6],user:6,userwarn:6,valu:0,variabl:0,version:[4,5],via:4,viewpoint:4,virtualenv:6,volunt:5,wa:7,wai:4,want:5,warn:6,we:4,web:5,welcom:[4,5],what:4,when:[4,5],which:[0,4],who:4,whoever:5,wiki:4,wish:0,within:[0,4],without:4,work:5,would:5,yang:7,year:0,you:[0,5,7],your:5,zcbp:0,zcta:0,zip:0},titles:["<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">community_ry2403.community_ry2403</span></code>","<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">community_ry2403</span></code>","API Reference","Changelog","Code of Conduct","Contributing","Example usage","community_ry2403"],titleterms:{"0":3,"1":3,"12":3,"2021":3,"class":0,"function":0,api:2,attribut:[0,4],bug:5,changelog:3,code:[4,5],community_ry2403:[0,1,7],conduct:[4,5],content:[0,1],contribut:[5,7],credit:7,document:5,enforc:4,exampl:6,featur:5,feedback:5,fix:5,get:5,guidelin:5,implement:5,instal:7,licens:7,modul:0,our:4,packag:1,pledg:4,pull:5,refer:2,report:5,request:5,respons:4,scope:4,standard:4,start:5,submit:5,submodul:1,type:5,usag:[6,7],v0:3,write:5}})

community_ry2403/src/community_ry2403/prepare.py
# scrape from Web, output CSV files, and save to the same path
import pandas as pd
import re

def get_estimate(col1):
    estimate = re.split('!!', col1, 1)[0].lower()
    return estimate

def get_label(col2):
    labels = re.split('!!', col2, 1)[1]
    return labels

def create_label(zbp_url='https://api.census.gov/data/2017/zbp/variables.html', \
                acs_url='https://api.census.gov/data/2019/acs/acs5/profile/variables.html'):
    """
    A function to output availble variables and labels from ZBP & ACS API with data scraped from Census websites.
    """
    var_bp = pd.read_html(zbp_url)[0]

    #create a new column differentiating absolute number or percent for converting report levels
    var_bp = var_bp[var_bp['Attributes'].notnull()][['Name','Label']].assign(estimate = 'estimate')
    var_bp = var_bp.reset_index().loc[0:9,:]
    var_bp.iat[-1,3] = 'percent'
    var_bp = var_bp.rename(columns={'Label':'labels'})

    var_acs = pd.read_html(acs_url)[0]
    var_acs = var_acs[var_acs['Attributes'].notnull()][['Name','Label']].reset_index().loc[0:1349,:]

    #use RegEx to capture the pattern identifying absolute number or percent for converting report levels     
    var_acs['labels'] = var_acs['Label'].apply(get_label)
    var_acs['estimate'] = var_acs['Label'].apply(get_estimate)

    var_acs = var_acs.iloc[:,var_acs.columns!='Label']

    label_csv = pd.concat([var_acs, var_bp]).drop_duplicates()

    return(label_csv)


def create_codebook(state_url='https://www.nrcs.usda.gov/wps/portal/nrcs/detail/?cid=nrcs143_013696' ,\
    zip_url='https://udsmapper.org/wp-content/uploads/2021/09/ZiptoZcta_Crosswalk_2021.xlsx'):
    """
    A function to create a crosswalk of ZIP-ZCTA-State Code with data scraped from some websites.
    """
    df = pd.read_html(state_url)[0]
    state = df.drop(df.tail(1).index).rename(columns={'Postal Code':'STATE','FIPS':'STATE_CODE'}) # drop last row

    crosswalk = pd.read_excel(zip_url, dtype = str).iloc[:,[0,2,4]].rename(columns={'ZIP_CODE':'ZIP'})

    codebook = pd.merge(crosswalk, state, on="STATE", how="left").iloc[:,[0,1,2,4]].dropna() 
    codebook['STATE_CODE'] = codebook['STATE_CODE'].astype(int)

    return(codebook)



community_ry2403/src/community_ry2403/community_ry2403.py
import os
import requests
import pandas as pd
import numpy as np
from fuzzywuzzy import fuzz

#------------------------------------------------------------------------------------- data prepared from web scraping
label_path = os.path.join(os.path.dirname(__file__), 'label_list.csv')
label_list = pd.read_csv(label_path)
codebook_path = os.path.join(os.path.dirname(__file__), 'codebook.csv')
codebook = pd.read_csv(codebook_path, dtype ='str')


#------------------------------------------------------------------------------------- functions prepared
# for fuzzy search
def get_ratio(str1,str2):
    return fuzz.token_set_ratio(str1,str2)


#prepare function:  convert report level (zip-zcta)
def get_code(area_code,level):
    """
    A function to get the crosswalk of ZCTA, ZIP, State, State Code, given a list of area codes.

    Parameters
    ----------
    area_code: a list of strings
        ZCTA or ZIP Code

    level: str
        ZIP or ZCTA; case-insensitive.

    Returns
    ----------
    A pandas dataframe containing the crosswalk of ZCTA, ZIP, State, State Code


    Examples
    ----------
    >>> get_code(area_code=['10025','10036'],level='zcta')

    [OUT] 
            ZIP	STATE	ZCTA	STATE_CODE
        3213	10025	NY	10025	36
        3224	10036	NY	10036	36
        3249	10108	NY	10036	36
        3250	10109	NY	10036	36
    """
    a = codebook.loc[ codebook[level.upper()].isin(area_code), :]
    return (a)


#prepare function: get business data from API
def business_data(year,variable,area_code,industry):
    """
    A function to get business data with from ZIP Code Business Pattern(ZCBP) API (ZIP Code level).

    Parameters
    ----------
    area_code: a list of strings
    ZIP code 
    year: int
    data reported year
    variable: a list of strings 
    variable codes of data

    Returns
    ----------
    A pandas dataframe containing the report level, the area code and values you looked up


    Examples
    ----------
    >>> v = ['EMP','EMP_N','ESTAB','PAYANN','PAYANN_N','PAYQTR1']
    >>> business_data(year=2019,variable=v,area_code=['10025','10036'],industry=72)

    [OUT] 
        EMP	EMP_N	ESTAB	PAYANN	PAYANN_N	PAYQTR1	NAICS2017	ZIP
        1	0	0	255	0	0	0	72	10025
        2	0	0	577	0	0	0	72	10036
    """
    try:
        dsource = 'zbp' if year <2018 else 'cbp'
        base_url = f'https://api.census.gov/data/{year}/{dsource}'

        zipcode = ','.join(area_code)
        cols = ','.join(variable)
        sector = str(industry)   #only allow single value

        search = requests.get(f'{base_url}?get={cols}&for=zipcode:{zipcode}&NAICS2017={sector}') 
        search.raise_for_status()
    except Exception as err:
        print(f'Other error occurred: {err}')
    else:
        df = pd.DataFrame(search.json())
        df = df.rename(columns=df.iloc[0]).drop(df.index[0]).rename(columns={'zip code':'ZIP'}) # make the first row be the header
        
        cols = df.columns.drop('ZIP')
        df[cols] = df[cols].apply(pd.to_numeric, errors='coerce')    # convert all columns except 'ZIP' to integer
        
        return(df)

# prepare function: get census data
def census_data(year,variable,area_code):
    """
    A function to get census data with from American Census Survey(ACS) API (ZCTA Code level).

    Parameters
    ----------
    area_code: a list of strings
    ZCTA code 
    year: int
    data reported year
    variable: a list of strings 
    variable codes of data

    Returns
    ----------
    A pandas dataframe containing the report level, the area code and values you looked up


    Examples
    ----------
    >>> v=['GEOCOMP','DP02_0001E','DP02_0002E','DP02_0003E']
    >>> census_data(2019,variable=v,area_code=['10025','10036'])

    [OUT] GEOCOMP	DP02_0001E	DP02_0002E	DP02_0003E	ZCTA
            1	0	17260	3111	719	    10025
            2	0	41355	14506	5309	10036
    """
    try:
        base_url = f'https://api.census.gov/data/{year}/acs/acs5/profile'

        zipcode = ','.join(area_code)
        cols = ','.join(variable)
        state = get_code(area_code,level='zcta')['STATE_CODE'].unique()[0]

        search = requests.get(f'{base_url}?get={cols}&for=zip%20code%20tabulation%20area:{zipcode}&in=state:{state}')
        search.raise_for_status()
    except Exception as err:
        print(f'Other error occurred: {err}')
    else:
        df = pd.DataFrame(search.json())
        df = df.rename(columns=df.iloc[0]).drop(df.index[0]).iloc[:,0:len(variable)]     # make the first row be the header
        df['ZCTA'] = area_code
        
        cols = df.columns.drop('ZCTA')
        df[cols] = df[cols].apply(pd.to_numeric, errors='coerce')    # convert all columns except 'ZIP' to integer

        return(df)


#------------------------------------------------------------------------------------- class1: find avaible variables with fuzzy search
class variables(object):
    def __init__(self):
        pass

    def find_variable(self,keyword):
        """
        a class to find available variables from American Census Survey(ACS), Zip Code Businiss Pattern(ZCBP) and Community Business Pattern(CBP) API.

        Parameters
        ----------
        keyword : str
        A string that you wish to look up 

        Returns
        ----------
        Pandas dataframe
        A number of variables that you may intereseted in with labels and variable codes

        Examples
        ----------
        >>> find_variable('total households marry people percent')

        [OUT]   variable	label
            0	DP02_0001E	HOUSEHOLDS BY TYPE!!Total households
            1	DP02_0001PE	HOUSEHOLDS BY TYPE!!Total households
            2	DP02_0014E	HOUSEHOLDS BY TYPE!!Total households!!Househol...
            3	DP02_0014PE	HOUSEHOLDS BY TYPE!!Total households!!Househol...
            4	DP02_0015E	HOUSEHOLDS BY TYPE!!Total households!!Househol...
            5	DP02_0015PE	HOUSEHOLDS BY TYPE!!Total households!!Househol...
            6	DP02PR_0001E	HOUSEHOLDS BY TYPE!!Total households
            7	DP02PR_0001PE	HOUSEHOLDS BY TYPE!!Total households
            8	DP02PR_0014E	HOUSEHOLDS BY TYPE!!Total households!!Househol...
            9	DP02PR_0014PE	HOUSEHOLDS BY TYPE!!Total households!!Househol...
            10	DP02PR_0015E	HOUSEHOLDS BY TYPE!!Total households!!Househol...
            11	DP02PR_0015PE	HOUSEHOLDS BY TYPE!!Total households!!Househol...
    
        """
        assert type(keyword)==str, 'keyword should be a string'
        
        df = pd.DataFrame(columns=['variable','label'])
        label = label_list['labels']
        variable = label_list['Name']
        n = len(label)
        for i,doc in enumerate(label):
            score = get_ratio(keyword,doc)
            new_row = pd.Series({'variable':variable[i],'label':doc})
            if score > 70:
                df = df.append(new_row,ignore_index=True)
        return(df)


#-------------------------------------------------------------------------------------class2: get census & business data with customized report level
class search(object):
    def __init__(self):
        pass

    def census(self,area_code,geography,year, variable):
        """
        A function to get community census data with customized report level (ZIP / ZCTA) from American Census Survey(ACS) API.
        If report level is ZCTA, keep the original value.
        If report level is ZIP, for percent data, use the original value within the correspondent ZCTA area, 
        and for absolute data, use the mean estimate within correspondent ZCTA area.

        Parameters
        ----------
        area_code: a list of strings
        ZIP code or ZCTA code
        geography: str
        data reported level: ZIP or ZCTA; case-insensitive.
        year: int
        data reported year
        variable: a list of strings 
        variable codes of data

        Returns
        ----------
        A pandas dataframe containing the report level, the area code and values you looked up


        Examples
        ----------
        >>> area_code = ['10025','10036']
        >>> geography = 'Zip'
        >>> year = 2019
        >>> variable = ['DP02_0001E','DP02_0001PE']
        >>> census(area_code,geography,year, variable)

        [OUT] ZIP	DP02_0001E_ZIP	DP02_0001PE_ZIP
            0	10025	17260.0	17260
            1	10036	13785.0	41355
        """
        assert all(isinstance(item, str) for item in area_code), "area_code should be a list of strings"
        assert geography.lower() in(['zip','zcta']), f"Data can only be reported by ZCTA or ZIP level, but geography is {geography}."
        assert type(year)==int,'year should be an integer'
        assert all(isinstance(item, str) for item in variable), "variable should be a list of strings"

        need_convert = (geography.lower()=='zip')  
        v = variable
        
        if need_convert == False:
            ans = census_data(year=year,variable=v,area_code = area_code)    
            
        else :
        
            est = label_list.loc[label_list['Name'].isin(v) ,:].reset_index()

            codemap = get_code(area_code,level=geography)[['ZCTA','ZIP']]
            conv_code = list(codemap['ZCTA'])

            df1 = get_code(conv_code,level='ZCTA')[['ZCTA','ZIP']]
            num = df1.groupby('ZCTA').count().rename(columns={'ZIP':'NUM'})
            df2 = census_data(year=year,variable=v,area_code = conv_code)

            merge_data = df1.merge(df2,on="ZCTA", how="left").merge(num,on="ZCTA", how="left")

            for i,var in enumerate(v):
                if est['estimate'][i] == 'percent':
                    merge_data[f'{var}_ZIP'] = merge_data[f'{var}']
                else:
                    merge_data[f'{var}_ZIP'] = merge_data[f'{var}']/merge_data['NUM']
            
            ans = merge_data.loc[merge_data['ZIP'].isin(area_code) , merge_data.columns.str.contains('ZIP')]
        
        return (ans)     



    def business(self,area_code,geography,year, variable,industry):
        """
        A function to get community business data with customized report level (ZIP / ZCTA) from Zip Code Businiss Pattern(ZCBP) and Community Business Pattern(CBP) API.
        If report level is ZIP, keep the original value.
        If report level is ZCTA, for percent data, use the mean estimate within the correspondent ZIP areas, 
        and for absolute data, use the sum estimate within correspondent ZIP areas.

        Parameters
        ----------
        area_code: a list of strings
            ZIP code or ZCTA code
        geography: str
            data reported level: ZIP or ZCTA; case-insensitive.
        year: int
            data reported year
        variable: a list of strings 
            variable codes of data
        industry: int
            NAICS 2017 Code identifying which industry of business data you wish to get

        Returns
        ----------
        A pandas dataframe containing the report level, the area code and values you looked up


        Examples
        ----------
        >>> area_code = ['79925','10025']
        >>> geography = 'Zcta'
        >>> year = 2019
        >>> variable = ['ESTAB','EMP_N','PAYQTR1_N']
        >>> industry=72
        >>> business(area_code, geography ,year, variable, industry)

        [OUT] ZCTA	PAYQTR1_N_ZCTA	EMP_N_ZCTA	ESTAB_ZCTA
            0	10025	0	0	255
            1	79925	0	0	214
        """
        assert all(isinstance(item, str) for item in area_code), "area_code should be a list of strings"
        assert geography.lower() in(['zip','zcta']), f"Data can only be reported by ZCTA or ZIP level, but geography is {geography}."
        assert type(year)==int,'year should be an integer'
        assert all(isinstance(item, str) for item in variable), "variable should be a list of strings"
        assert type(industry)==int,'industry code should be an integer'

        need_convert = (geography.lower()=='zcta')  
        v = variable
        
        if need_convert == False:
            ans = business_data(year=year,variable=v,area_code = area_code,industry=industry)    
            
        else :
            est = label_list.loc[label_list['Name'].isin(v) ,:].reset_index()

            df1 = get_code(area_code,level='ZCTA')[['ZCTA','ZIP']]
            conv_code = list(df1['ZIP'])
            df2 = business_data(year=year,variable=v,area_code=conv_code,industry=industry)
            merge_data = df1.merge(df2,on="ZIP", how="right")

            a = merge_data.loc[:,merge_data.columns!='ZIP']
            ans = merge_data[['ZCTA']]

            for i,var in enumerate(v):
                if est['estimate'][i] == 'percent':
                    t = a.groupby(['ZCTA'])[[f'{var}']].mean().rename(columns={f'{var}':f'{var}_ZCTA'})
                else:
                    t = a.groupby(['ZCTA'])[[f'{var}']].sum().rename(columns={f'{var}':f'{var}_ZCTA'})
                ans = t.merge(ans,on='ZCTA',how='left')
        return(ans)